File: .snippets/code/polkadotXcm/xcmExecute/web3.py
from web3 import Web3

abi = 'XCM-UTILS-ABI-HERE'  # Paste or import the x-tokens ABI
# This is for demo purposes, never store your private key in plain text
privateKey = 'INSERT-YOUR-PRIVATE-KEY'
# The wallet address that corresponds to your private key
address = 'INSERT-YOUR-ADDRESS'

## Create Web3 provider ##
web3 = Web3(Web3.HTTPProvider('https://rpc.api.moonbase.moonbeam.network'))

## Create contract instance of the XCM Utilities Precompile ##
xcmUtils = web3.eth.contract(
    # XCM Utilities Precompile address
    address='0x000000000000000000000000000000000000080C',
    abi=abi
)


def executeXcmMessageLocally():
    ## Define parameters required for the xcmExecute function ##
    encodedCalldata = 'INSERT-ENCODED-CALLDATA'
    maxWeight = 100000000000

    ## Execute the custom XCM message ##
    # Craft the extrinsic
    tx = xcmUtils.functions.xcmExecute(
        encodedCalldata,
        maxWeight
    ).buildTransaction(
        {
            'from': address,
            'nonce': web3.eth.get_transaction_count(address),
        }
    )
    # Sign transaction
    signedTx = web3.eth.account.sign_transaction(tx, privateKey)
    # Send tx
    hash = web3.eth.send_raw_transaction(signedTx.rawTransaction)
    receipt = web3.eth.wait_for_transaction_receipt(hash)
    print(f'Transaction receipt: { receipt.transactionHash.hex() }')


executeXcmMessageLocally()

File: .snippets/code/polkadotXcm/xcmExecute/web3.js
import ABI from './xcmUtilsABI.js'; // Import the XCM Utilities Precompile ABI
import Web3 from 'web3'; // Import Web3 library

const privateKey = 'INSERT_PRIVATE_KEY';
const xcmUtilsAddress = '0x000000000000000000000000000000000000080C';

/* Create Web3 provider */
const web3 = new Web3('https://rpc.api.moonbase.moonbeam.network'); // Change to network of choice

/* Create contract instance of the XCM Utilities Precompile */
const xcmUtils = new web3.eth.Contract(
  ABI,
  xcmUtilsAddress,
  { from: web3.eth.accounts.privateKeyToAccount(privateKey).address } // 'from' is necessary for gas estimation
);

const executeXcmMessageLocally = async () => {
  /* Define parameters required for the xcmExecute function */
  const encodedCalldata = 'INSERT_ENCODED_CALLDATA';
  const maxWeight = '400000000';

  /* Send the custom XCM message */
  // Craft the extrinsic
  const tx = await xcmUtils.methods.xcmExecute(encodedCalldata, maxWeight);
  // Sign transaction
  const signedTx = await web3.eth.accounts.signTransaction(
    {
      to: xcmUtilsAddress,
      data: tx.encodeABI(),
      gas: await tx.estimateGas(),
    },
    privateKey
  );
  // Send the signed transaction
  const sendTx = await web3.eth.sendSignedTransaction(signedTx.rawTransaction);
  console.log(`Transaction receipt: ${sendTx.transactionHash}`);
};

executeXcmMessageLocally();
